local/modules/TheliaSmarty/Template/Plugins/UrlGenerator.php

```diff
@@ -119,6 +122,36 @@ public function generateUrlFunction($params, &$smarty)
                 $this->getArgsFromParam($params, array_merge(['noamp', 'path', 'file', 'target'], $excludeParams)),
                 $mode
             );
+            
+            $request = $this->getRequest();
+            $requestedLangCodeOrLocale = $params["lang"];
+            $view = $request->attributes->get('_view', null);
+            $viewId = $view === null ? null : $request->query->get($view . '_id', null);
+
+            if (null !== $requestedLangCodeOrLocale) {
+                if (strlen($requestedLangCodeOrLocale) > 2) {
+                    $lang = LangQuery::create()->findOneByLocale($requestedLangCodeOrLocale);
+                } else {
+                    $lang = LangQuery::create()->findOneByCode($requestedLangCodeOrLocale);
+                }
+                $urlRewrite = RewritingUrlQuery::create()
```

Comment:
You can move this request inside the ` if (ConfigQuery::isMultiDomainActivated()) {` test to save on request